File: document_processing/pipeline_modules/doc_detector/doc_detector.py
from ..base_module import BaseModule
from .image_transformation import fix_perspective, rectify_pages, stitch_pages
from typing import Union
from pathlib import Path
import cv2
import numpy as np

# A second document page (passport spread) must cover at least this fraction of
# the largest segment's area to be kept. Confidence is NOT a reliable filter
# here (spurious strips often score 0.9+ while faint real pages score ~0.65),
# so selection is area-based. Genuine spread pages run ~0.84-1.0; spurious
# background blobs run <=~0.5, so 0.6 separates them.
SECOND_SEGMENT_AREA_FRAC = 0.6

# A segment with (almost) no ink inside it is not a document page, however
# confident the model is: the white lid of a flatbed scanner next to a
# passport scores 0.90-0.96 as 'Document' and, being 3x larger than a page,
# used to win the area-based selection above and push both real pages out
# (~20 of 100 scans in the Damir set came out as an empty canvas). Ink is the
# mean gradient magnitude inside the eroded mask at ~400 px: measured 1.5-5.4
# on lids, 19-101 on passport pages (19 = a badly blurred one), 24-43 on the
# mostly bare registration page. A blank segment is dropped only when another
# segment with real ink exists, so a lone blank sheet still goes through as
# before.
BLANK_INK = 8.0
INK_SCALE_PX = 400


def segment_ink(gray: np.ndarray, contour) -> float:
    """Mean gradient magnitude inside ``contour`` (eroded so the segment's own
    edge does not count), on the image downscaled to INK_SCALE_PX."""
    h, w = gray.shape[:2]
    s = INK_SCALE_PX / float(max(h, w))
    small = cv2.resize(gray, (max(1, int(w * s)), max(1, int(h * s))), interpolation=cv2.INTER_AREA)
    pts = np.asarray(contour, dtype=np.float32).reshape(-1, 2)
    if len(pts) < 3:
        return 0.0
    m = np.zeros(small.shape, np.uint8)
    cv2.fillPoly(m, [np.int32(np.round(pts * s))], 255)
    m = cv2.erode(m, np.ones((5, 5), np.uint8))
    if not m.any():
        return 0.0
    gx = cv2.Sobel(small, cv2.CV_32F, 1, 0, ksize=3)
    gy = cv2.Sobel(small, cv2.CV_32F, 0, 1, ksize=3)
    return float(np.hypot(gx, gy)[m > 0].mean())


def drop_blank_segments(img: np.ndarray, segm):
    """Indices of the segments to keep: blank ones (ink < BLANK_INK) are
    dropped when at least one inked segment exists. Returns the kept
    indices and the per-segment ink values."""
    gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY) if img.ndim == 3 else img
    ink = [segment_ink(gray, s) if s is not None else 0.0 for s in segm]
    if any(v >= BLANK_INK for v in ink):
        keep = [i for i, v in enumerate(ink) if v >= BLANK_INK]
    else:
        keep = list(range(len(segm)))
    return keep, ink



class DocDetector(BaseModule):
    """Detects document and fixes perspective issues.

    Detects the document quadrangle, segments it from the background,
    and fixes perspective issues by transforming the document to
    a rectangular shape.

    Provides options to get just the detection outputs or also warp
    the document image to fix its perspective.

    """
    def __init__(self, model_format: str = 'ONNX', device='cpu', verbose: bool=False, runtime: str = None):
        """Initializes the document detection model"""
        self.model_name = 'DocDetector'
        super().__init__(self.model_name, model_format=model_format, device=device, verbose=verbose, runtime=runtime)

    def predict(self, img: Union[str, Path, np.ndarray]) -> dict:
        """Predicts document detection outputs.

        Args:
            img: Input document image

        Returns:
            Dictionary with bboxes, masks and segmentation
        """
        img = self.load_img(img)

        bbox, mask, segm = self.model.predict(img)
        meta = {
            self.model_name:
                {
                    'bbox': bbox,
                    'mask': mask,
                    'segm': segm,
                }
        }
        return meta

    def predict_transform(self, img: Union[str, Path, np.ndarray], stack: str = 'auto',
                          max_pages: int = 2) -> dict:
        """Predicts outputs and fixes document perspective.

        Args:
            img: Input document image
            stack: Multi-document merge direction passed to fix_perspective
                   ('auto' picks horizontal/vertical from page layout).
            max_pages: Max document segments to keep. Single-page doc types
                   should pass 1 so background blobs are never stitched in;
                   internal-passport spreads pass 2.

        Returns:
            Dictionary with detections outputs and warped image
        """
        img = self.load_img(img)
        bbox, mask, segm = self.model.predict(img)

        if segm:

            # Select document segments by contour area: keep the largest page
            # plus, optionally, a second page whose area is at least
            # SECOND_SEGMENT_AREA_FRAC of it. This drops spurious thin
            # fragments (which can still carry high confidence) while keeping
            # genuine two-page spreads. Capped at 2 pages.
            # First drop segments without ink (a scanner lid, a blank sheet
            # next to the document): see BLANK_INK. Then the area rule.
            inked, ink_values = drop_blank_segments(img, segm)
            if len(inked) < len(segm):
                bbox = [bbox[i] for i in inked]
                mask = [mask[i] for i in inked]
                segm = [segm[i] for i in inked]
            areas = []
            for s in segm:
                pts = np.asarray(s, dtype=np.float32).reshape(-1, 2) if s is not None else np.empty((0, 2))
                areas.append(cv2.contourArea(pts) if len(pts) >= 3 else 0.0)

            if areas and max(areas) > 0:
                order = list(np.argsort(areas)[::-1])
                max_area = areas[order[0]]
                keep = [order[0]]
                for idx in order[1:]:
                    if len(keep) >= max(1, max_pages):
                        break
                    if areas[idx] >= SECOND_SEGMENT_AREA_FRAC * max_area:
                        keep.append(idx)
                keep = sorted(keep)
                bbox = [bbox[i] for i in keep]
                mask = [mask[i] for i in keep]
                segm = [segm[i] for i in keep]


            try:
                # Pages are rectified separately and stitched afterwards, so
                # the individual pages can be handed to the pipeline: a
                # detector's 640x640 input gives a stitched spread only half of
                # itself per page. The stitched canvas is still produced here -
                # it stays the thing every existing consumer expects.
                pages, quads, borders_img = rectify_pages(img=img, segments=segm)
                if pages:
                    result_img, placements = stitch_pages(pages, quads, stack=stack)
                else:
                    result_img, placements = img, []
            except Exception as e:
                print(f'[!] Failed to fix perspective: {e!r}')
                result_img = borders_img = img
                pages, quads, placements = [], [], []
        else:
            result_img = borders_img = img
            pages, quads, placements = [], [], []
        meta = {
            self.model_name:
                {
                    'bbox': bbox,
                    'mask': mask,
                    'segm': segm,
                    'border_img': borders_img,
                    'warped_img': result_img,
                    # per-page rectified images plus the quad each came from and
                    # where it landed on the stitched canvas; empty for a
                    # single-page document, where 'warped_img' already IS the page
                    'pages': pages,
                    'page_quads': quads,
                    'page_placements': placements,

                }
        }

        return meta

File: ports/go/internal/docproc/modules/docdetector.go
package modules

import (
	"fmt"
	"sort"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/config"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/imaging"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/inference"
	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/models"
)

// secondSegmentAreaFrac is the share of the largest page's area a second segment must
// reach to be kept.
//
// Selection is by AREA and not by confidence, and the reference explains why:
// spurious thin strips routinely score 0.9+ while a faint genuine page can score ~0.65,
// so confidence is not a usable filter here. Real spread pages run ~0.84-1.0 of the
// largest and background blobs stay under ~0.5, which is what 0.6 separates.
const secondSegmentAreaFrac = 0.6

// blankInk is BLANK_INK (doc_detector.py:15-26). A segment with (almost) no ink inside
// it is not a document page, however confident the model is: the white lid of a flatbed
// scanner next to a passport scores 0.90-0.96 as 'Document' and, being 3x larger than a
// page, used to win the area-based selection and push both real pages out (~20 of 100
// scans in the Damir set came out as an empty canvas). Ink is the mean gradient
// magnitude inside the eroded mask at ~400 px: measured 1.5-5.4 on lids, 19-101 on
// passport pages (19 = a badly blurred one), 24-43 on the mostly bare registration
// page. A blank segment is dropped only when another segment with real ink exists, so a
// lone blank sheet still goes through as before.
const blankInk = 8.0

// DocDetector finds the document's borders and rectifies its perspective.
// Port of pipeline_modules/doc_detector. Uses the Borders artifact.
type DocDetector struct {
	model *models.SegmentationModel
}

func NewDocDetector(root string, paths *config.ModelPaths, format string,
	device inference.Device, threads int) (*DocDetector, error) {

	dir, err := paths.Dir("DocDetector", format)
	if err != nil {
		return nil, err
	}
	m, err := models.LoadSegmentation(root, dir, device, threads)
	if err != nil {
		return nil, fmt.Errorf("modules: DocDetector: %w", err)
	}
	return &DocDetector{model: m}, nil
}

func (d *DocDetector) Close() error { return d.model.Close() }

// DocDetectorResult is PredictTransform's output: the stitched canvas plus, for a
// multi-page spread, the individual rectified pages, the (expanded, clipped) quad each
// came from, and where each landed on the canvas - exactly what DocDetector.predict_transform
// hands to the pipeline (pages/page_quads/page_placements), so the caller can run field
// detection PER PAGE (pipeline.py's _fields_from_pages) instead of on the half-size
// letterboxed spread. Pages is nil/empty for a single-page document; the CALLER owns
// every Image here (Canvas and each element of Pages), and Pages Images are separate
// Mats from Canvas even when len(Pages)==1 (StitchPages returns that one page AS the
// canvas in that case, i.e. Pages[0] and Canvas alias the same Mat then - Close only
// once).
type DocDetectorResult struct {
	Canvas      imaging.Image
	Segments    [][]imaging.Point
	Pages       []imaging.Image
	PageQuads   [][]imaging.Point
	Placements  []imaging.PagePlacement
}

// PredictTransform returns the perspective-corrected canvas.
//
// maxPages caps how many document segments are kept: 1 for single-page types so a
// background blob can never be stitched in, 2 for an internal-passport spread.
//
// When no usable segment is found the ORIGINAL image is returned. That is not a
// fallback bolted on for safety — it is what the reference does, and a port that
// errored instead would fail every document whose borders the model cannot see.
// Returns the SELECTED contours alongside the canvas, so the conformance harness can
// compare them (the borders.segments stage) and localise a divergence to the mask
// rather than to the warp. Nil when border detection found nothing.
func (d *DocDetector) PredictTransform(img imaging.Image, maxPages int) (DocDetectorResult, error) {
	_, segments, err := d.model.Predict(img)
	if err != nil {
		return DocDetectorResult{}, err
	}
	if len(segments) == 0 {
		return DocDetectorResult{Canvas: img.Clone()}, nil
	}

	// First drop segments without ink (a scanner lid, a blank sheet next to the
	// document): see blankInk. Then the area rule (doc_detector.py:125-131).
	inked := dropBlankSegments(img, segments)
	if len(inked) < len(segments) {
		filtered := make([][]imaging.Point, 0, len(inked))
		for _, i := range inked {
			filtered = append(filtered, segments[i])
		}
		segments = filtered
	}

	kept := selectPages(segments, maxPages)
	if len(kept) == 0 {
		return DocDetectorResult{Canvas: img.Clone()}, nil
	}

	chosen := make([][]imaging.Point, 0, len(kept))
	for _, i := range kept {
		chosen = append(chosen, segments[i])
	}

	// Pages are rectified separately and stitched afterwards (matching the reference's
	// doc_detector.predict_transform), so a two-page spread can be handed to the field
	// detector one page at a time - a 640x640 detector input gives a stitched spread
	// only half of itself per page.
	pages, quads := imaging.RectifyPages(img, chosen, imaging.DocMarginFrac)
	if len(pages) == 0 {
		return DocDetectorResult{Canvas: img.Clone(), Segments: chosen}, nil
	}
	if len(pages) < 2 {
		// One page: StitchPages returns it AS the canvas (no copy needed - nothing
		// downstream needs Pages for fewer than 2 pages anyway, both consumers gate
		// on len(Pages) >= 2), so hand it the original directly.
		canvas, _, ok := imaging.StitchPages(pages, quads, imaging.StackAuto)
		if !ok {
			return DocDetectorResult{Canvas: img.Clone(), Segments: chosen}, nil
		}
		return DocDetectorResult{Canvas: canvas, Segments: chosen}, nil
	}
	// 2+ pages: StitchPages CONSUMES (Closes) every page it is handed - but the
	// per-page field detector (pipeline._fields_from_pages) needs the pages
	// themselves to survive past this call, so it is handed CLONES, never the
	// originals. Skipping this clone is a double-free: Pages and the canvas each
	// point at freed memory the instant this returns (caught the hard way, see the
	// task log - this exact bug crashed every internal-passport document).
	stitchInput := make([]imaging.Image, len(pages))
	for i, p := range pages {
		stitchInput[i] = p.Clone()
	}
	canvas, placements, ok := imaging.StitchPages(stitchInput, quads, imaging.StackAuto)
	if !ok {
		closeImages(pages)
		return DocDetectorResult{Canvas: img.Clone(), Segments: chosen}, nil
	}
	return DocDetectorResult{
		Canvas: canvas, Segments: chosen,
		Pages: pages, PageQuads: quads, Placements: placements,
	}, nil
}

func closeImages(imgs []imaging.Image) {
	for _, im := range imgs {
		_ = im.Close()
	}
}

// dropBlankSegments returns the indices of the segments to keep: blank ones
// (ink < blankInk) are dropped when at least one inked segment exists.
// Port of doc_detector.drop_blank_segments (doc_detector.py:48-58).
func dropBlankSegments(img imaging.Image, segments [][]imaging.Point) []int {
	ink := make([]float64, len(segments))
	anyInked := false
	for i, s := range segments {
		ink[i] = imaging.SegmentInk(img, s)
		if ink[i] >= blankInk {
			anyInked = true
		}
	}
	keep := make([]int, 0, len(segments))
	for i := range segments {
		if !anyInked || ink[i] >= blankInk {
			keep = append(keep, i)
		}
	}
	return keep
}

// selectPages ranks segments by contour area and applies the area-fraction rule.
//
// Returns indices in ASCENDING order, matching the reference's `sorted(keep)`, because
// the order then decides which page FixPerspective treats as first when stitching a
// spread.
func selectPages(segments [][]imaging.Point, maxPages int) []int {
	areas := make([]float64, len(segments))
	for i, s := range segments {
		if len(s) >= 3 {
			areas[i] = contourArea(s)
		}
	}

	order := make([]int, len(areas))
	for i := range order {
		order[i] = i
	}
	// Descending by area, stable so equal areas keep their original relative order.
	sort.SliceStable(order, func(a, b int) bool { return areas[order[a]] > areas[order[b]] })

	if len(order) == 0 || areas[order[0]] <= 0 {
		return nil
	}
	limit := maxPages
	if limit < 1 {
		limit = 1
	}
	maxArea := areas[order[0]]
	keep := []int{order[0]}
	for _, idx := range order[1:] {
		if len(keep) >= limit {
			break
		}
		if areas[idx] >= secondSegmentAreaFrac*maxArea {
			keep = append(keep, idx)
		}
	}
	sort.Ints(keep)
	return keep
}

// contourArea is the shoelace formula, matching cv2.contourArea's magnitude.
//
// OpenCV returns the absolute area for a simple polygon, so the sign of the traversal
// direction is discarded — which is what the reference's comparisons assume.
func contourArea(pts []imaging.Point) float64 {
	n := len(pts)
	if n < 3 {
		return 0
	}
	var acc float64
	for i := 0; i < n; i++ {
		j := (i + 1) % n
		acc += pts[i].X*pts[j].Y - pts[j].X*pts[i].Y
	}
	if acc < 0 {
		acc = -acc
	}
	return acc / 2
}
